File: scripts/BordaCountFeatures.R
suppressPackageStartupMessages(library(readr))
suppressPackageStartupMessages(library(dplyr))

inFilePath <- commandArgs()[7]
outFilePath <- commandArgs()[8]
numTop <- commandArgs()[9]

buildRankMatrix <- function(dataSubset)
{
  featuresMatrix <- t(data.frame(lapply(dataSubset$Features, function(x) {strsplit(x,',')})))
  rownames(featuresMatrix) <- NULL

  return(featuresMatrix)
}

rankAggregate <- function(featuresMatrix)
{
  uniqueFeatures <- unique(as.vector(featuresMatrix))
  rankMatrix <- NULL

  for (i in 1:nrow(featuresMatrix))
  {
    rowFeatures <- featuresMatrix[i,]
    rowRanks <- match(uniqueFeatures, rowFeatures)
    rankMatrix <- cbind(rankMatrix, rowRanks)
  }

  meanRank <- apply(rankMatrix, 1, mean)

  result <- data.frame(Feature=uniqueFeatures, Mean_Rank=meanRank)
  return(result[order(meanRank),])
}

#Description	Iteration	Algorithm	Features
#Description	Iteration	Fold	Algorithm	Features
#Description	Iteration	Ensemble_Algorithm	Algorithm	Features
#Description	Iteration	Fold	Ensemble_Algorithm	Algorithm	Features
data <- suppressMessages(read_tsv(inFilePath))

if ("ERROR" %in% data$Features)
{
  cat("\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n")
  cat("Error: Feature ranks could not be summarized because at least one individual algorithm experienced an error. To troubleshoot the error, reexecute ShinyLearner in verbose mode.\n")
  cat("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n\n")
  stop()
}

#groupRankMatrix <- NULL
#for (algorithm in unique(data$Algorithm))
#{
#  algorithmData <- filter(data, Algorithm==algorithm)
#  algorithmRanks <- rankAggregate(buildRankMatrix(algorithmData))
#  groupRankMatrix <- rbind(groupRankMatrix, as.character(algorithmRanks$Features))
#}

#overallRanks <- rankAggregate(groupRankMatrix)

overallRanks <- rankAggregate(buildRankMatrix(data))

if (!is.na(numTop))
{
  numTop <- min(c(as.integer(numTop), nrow(overallRanks)))
  overallRanks <- overallRanks[1:numTop,]
}

write_tsv(overallRanks, outFilePath)

File: scripts/BordaCountFeatures.py
import sys
from operator import itemgetter, attrgetter

inFilePath = sys.argv[1]
outFilePath = sys.argv[2]

#Description Outer_Iteration Inner_Iteration Algorithm   Features
inFile = open(inFilePath)
headerItems = inFile.readline().rstrip("\n").split("\t")

rankDict = {}

for line in inFile:
    lineItems = line.rstrip("\n").split("\t")
    features = lineItems[-1]

    if features == "ERROR":
        print("\n!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
        print("Error: Feature ranks could not be summarized because at least one individual algorithm experienced an error. To troubleshoot the error, reexecute the analysis in verbose mode.")
        print("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n")
        sys.exit(1)

    features = features.split(",")
    algorithm = lineItems[-2]

    if not algorithm in rankDict:
        rankDict[algorithm] = {}

    # Populate the dictionary with ranks for each iteration
    for i in range(len(features)):
        rank = i + 1
        feature = features[i]

        if feature in rankDict[algorithm]:
            rankDict[algorithm][feature].append(float(rank))
        else:
            rankDict[algorithm][feature] = [float(rank)]

inFile.close()

outFile = open(outFilePath, 'w')
outFile.write("Feature_Selection_Algorithm\tFeature\tMean_Rank\n")

for algorithm, algoRankDict in sorted(rankDict.items()):
    meanRanks = []
    for feature, ranks in algoRankDict.items():
        meanRank = sum(ranks) / len(ranks)
        meanRanks.append([feature, meanRank])

    meanRanks.sort(key=itemgetter(1))

    for x in meanRanks:
        outFile.write("{}\t{}\t{:.1f}\n".format(algorithm, x[0], x[1]))

outFile.close()
